src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java

```diff
@@ -371,42 +363,50 @@ private void rescaleIfNeeded(long now)
             {
                 try
                 {
-                    rescale(now);
+                    decayingBuckets = rescale(now);
                 }
                 finally
                 {
-                    decayLandmark = now;
                     rescaling.set(false);
                 }
             }
         }
     }
 
-    private void rescale(long now)
+    /**
+     * rescale races with update;
+     * If rescale is called at the same time as an update, and an update falls into an already
+     * processed bucket this update will be lost. We accept that for the sake of update performance.
+     * OTOH, if the update falls into a bucket that is yet to be processed, it will be added
+     * with a weight that is consistent with the rescaleFactor used by rescale, so the new rescaled
+     * buckets will all have values consistent with the new decay landmark.
+     */
+    private DecayingBuckets rescale(long now)
     {
-        // despite striping its safe to rescale each bucket individually
-        final double rescaleFactor = forwardDecayWeight(now);
-        for (int i = 0; i < decayingBuckets.length(); i++)
+        final double rescaleFactor = decayingBuckets.forwardDecayWeight(now);
```

Comment:
We are reading the `decayingBuckets` three times here and if the `rebase` method is called, this could potentially cause a problem. So I would save the decayingBuckets to a local ref and read it.